### DISCUSSION TITLE
Verbindung zwischen TTN und Stadtpuls

### DISCUSSION THREAD

**Discussion by JerreyGlen** [technologiestiftung/stadtpuls#8] [Q&A] [answered]
Hello everyone,
Es soll eine Plattform zur Visualisierung der Umweltdaten an der HTW Berlin und auf Satdtpuls. WIr würden jetzt unsere Daten (auf TTN) auf der Stadtpuls Seite darstellen und wir sind auf ein Paar Schwierigkeiten gestoßen.
Bis jetzt werden Keine unserer Messwerte  auf der Stadtpuls Seite gezeigt. Es gibt Fehler bei Senden des Webhooks auf TTN. Unserer payload hat 3 Werte statt 4.
Als End-device auf der TTN haben wir unserer ESP 32 LoRa registriert und nicht jeder Sensor (Es wird nur der erste parameter gezeigt) und auf Stadtpuls wurde ein Luftfeuchtigkeitssensor (Siehe Anhang) registriert. Was kann die Ursache sein, warum der Webhook nicht gesendet wird ? Noch dazu brauchen wir Typen von allen Sensoren von einem bestimmten Payload. 
Anders gesagt, wollen wir alle Sensortypen und -daten an einem bestimmten Standort anzeigen und die JSON-datei enthält nur Werte. Auf Stadtpuls wird nur den ersten Typ dargestellt.
WIe Können wir allen Sensortypen haben? wenn es schwierig oder unmöglich ist, dann ist es kein Problem.

**Comment by ff6347** [comment]
> Bis jetzt werden Keine unserer Messwerte auf der Stadtpuls Seite gezeigt.

Can you give me a username so I can investigate?

> Es gibt Fehler bei Senden des Webhooks auf TTN. 

The TTN console should show you the error mesage. Can you tell me what  it is?


> Unserer payload hat 3 Werte statt 4. 

Should not be a problem.


> Als End-device auf der TTN haben wir unserer ESP 32 LoRa registriert und nicht jeder Sensor (Es wird nur der erste parameter gezeigt) und auf Stadtpuls wurde ein Luftfeuchtigkeitssensor (Siehe Anhang) registriert.

There is no attachment. And yes we only display the first value. We are working on a better display of multiple values.

> Was kann die Ursache sein, warum der Webhook nicht gesendet wird ? 

There might be many problems, wrong url, wrong payload formatter, error in the auth token…

>Noch dazu brauchen wir Typen von allen Sensoren von einem bestimmten Payload. Anders gesagt, wollen wir alle Sensortypen und -daten an einem bestimmten Standort anzeigen und die JSON-datei enthält nur Werte. Auf Stadtpuls wird nur den ersten Typ dargestellt. WIe Können wir allen Sensortypen haben? wenn es schwierig oder unmöglich ist, dann ist es kein Problem.

I don't get the last part. Would english be more suitable for you?

**Reply by ff6347** [reply]
I would guess you are orock and the sensor is this one? https://stadtpuls.com/sensors/58

**Reply by JerreyGlen** [reply]
Exactly, i am Orock

**Reply by JerreyGlen** [reply]
you already answered the last question, thanks.

**Reply by ff6347** [reply]
To debug this it would be nice to see:

- your payload formatter
- an error message from the TTN console 
- maybe a screenshot from your webhook (make sure to blacken the auth token)

Can you confirm that your data reaches the TTN console?

**Reply by JerreyGlen** [reply]
Yes, you can see it on screenshots.

**Comment by JerreyGlen** [comment]
The error message is: Failed to a Webhook. 
![Bild_2021-12-21_231719](https://user-images.githubusercontent.com/75827115/147247687-b57f3ebf-0081-426f-bbcc-6887f11052ac.png)
<img width="917" alt="Screenshot_2021-12-21_231330" src="https://user-images.githubusercontent.com/75827115/147247690-a3b9db7e-66ba-4e2e-9ecc-6701225f3ee6.png">
<img width="905" alt="Screenshot_2021-12-21_231439" src="https://user-images.githubusercontent.com/75827115/147247693-6397a843-e332-430d-81b6-0f5089a8ec4d.png">
<img width="628" alt="Screenshot_2021-12-21_225743" src="https://user-images.githubusercontent.com/75827115/147247696-912bd3ee-89dd-4f8f-a02c-567d044925f2.png">
<img width="587" alt="Screenshot_2021-12-21_225825" src="https://user-images.githubusercontent.com/75827115/147247699-c232fe13-3382-4622-a27b-1e7a442373b5.png">
<img width="946" alt="Screenshot_2021-12-21_230228" src="https://user-images.githubusercontent.com/75827115/147247702-7af5a312-2872-498e-b9e6-88d1fb181cb1.png">
<img width="943" alt="Screenshot_2021-12-21_230250" src="https://user-images.githubusercontent.com/75827115/147247703-8baeebed-3029-421e-bcf1-98e5fd8a900e.png">

**Reply by ff6347** [reply]
Payload formatter looks correct.

One error could be that the base url ends with a `/` and the endpoint starts with a`/`. Means this is a `https://api.stadtpuls.com//api/v2/integrations/ttn/v3` which will create an error

Can you expand the error that says failed to send webhook? Is there somewhere an error mesage? Normally our api should send something in response. 404 not found or 401 not authorized or 400 wrong payload and what not.

I'll be afk for a few hours. I'll try to check later on.

**Comment by JerreyGlen** [comment]
![TTN2](https://user-images.githubusercontent.com/75827115/147263822-390de9cb-a097-4f94-8668-5dbe5986d800.JPG)
![TTN1](https://user-images.githubusercontent.com/75827115/147263826-6c867e67-9092-46de-9d82-214df5998046.JPG)

**Comment by ff6347** [comment]
Hi @JerreyGlen I see the issue. The error is 401. Means you are not authorized.

I looked up your tokens. From the screenshot it looks like you added the description of your token and not the token itself. If you don't have it anymore, just create a new one and copy it from the modal. Remember. You will see the token only once. Then it is up to you to store it somewhere safe.

**Reply by JerreyGlen** [reply]
Hi @ff6347,
I understood your message and tried to create a new token named Token 1. I use the very long key which you just see once and that was not useful. It would be nice to have a Videoconference and begin the whole process together. When would you have time?

**Reply by ff6347** [reply]
Hi Jerry,

Currently I'm on vacation. I'll be back in the office on the third of January 2022. I would rather prefer to do this here in text form so others can learn from these problems then doing one on one phone/videocall support. The second is black boxing information. 

Please read the docs again and let us know where to are struggling. Maybe in a new thread specific to the docs section?


Also:

1. Remove the slash from your base URL 
2. Add the token as `authentication` header field with the word Bearer in front of it. `Bearer YOURTOKENHERE`
3. If an error occurs again copy and paste the error message from the TTN console to this thread. 
4. Inspect the HTTP errors 400 means faulty request, 401 means not authorized, 404 means not found.

**Reply by JerreyGlen** [reply]
Could not create webhook
invalid `headers[authentication]`: value length must be at most 256 runes

{
  "code": 3,
  "message": "error:pkg/errors:validation (invalid `webhook`: embedded message failed validation)",
  "details": [
    {
      "@type": "type.googleapis.com/ttn.lorawan.v3.ErrorDetails",
      "namespace": "pkg/errors",
      "name": "validation",
      "message_format": "invalid `{field}`: {reason}",
      "attributes": {
        "field": "webhook",
        "name": "SetApplicationWebhookRequestValidationError",
        "reason": "embedded message failed validation"
      },
      "correlation_id": "717faddaf3e5411085f806856dc2e664",
      "cause": {
        "namespace": "pkg/errors",
        "name": "validation",
        "message_format": "invalid `{field}`: {reason}",
        "attributes": {
          "field": "headers[authentication]",
          "name": "ApplicationWebhookValidationError",
          "reason": "value length must be at most 256 runes"
        },
        "correlation_id": "b839a784bfa1468a84f0b0e232f62f02",
        "code": 3
      },
      "code": 3
    }
  ]
}

**Reply by JerreyGlen** [reply]
i used my account mekta82 on Stadtpuls and always used the account of Orock on TTN. I was adding a webhook and now i am seeing the same message. The token is too long. It is really easy to apply all what we can see on the Stadtpuls webpage.
What can I do?

**Reply by ff6347** [reply]
Hi Jerry, I can reproduce the error. That is new. We had no issues with our tokens until now. They are normal JWT tokens and there should be no char limit on headers values. I will investigate further. Thanks for the report. Seems to be a ttn related issue.

**Reply by JerreyGlen** [reply]
Hi @ff6347,
it is ttn related and it is in relation with stadtpuls. How much time should it take to be solved?
Can I help?

**Reply by ff6347** [reply]
> Hi @ff6347,
> 
> it is ttn related and it is in relation with stadtpuls. How much time should it take to be solved?
> 
> Can I help?
> 
>  

We are not related at all and have no influence. I'll start today to find out today where this problem is coming form

**Reply by ff6347** [reply]
@JerreyGlen I created a bug report on the TheThingsNetwork/lorawan-stack repo. I hope this will be fixed soon. We will work on a workaround until then. Subscribe to that issue if you want to stay up to date. https://github.com/TheThingsNetwork/lorawan-stack/issues/5062

**Reply by ff6347** [reply]
We have a fix and should release a new version tomorrow https://github.com/technologiestiftung/stadtpuls-api/pull/111

**Comment by ff6347** [comment, marked as answer]
@JerreyGlen We updated our API. All newly generated tokens should work again.

**Reply by JerreyGlen** [reply]
Ok very Nice, I try now another time.

**Reply by JerreyGlen** [reply]
I received a new error [400.]
![new error](https://user-images.githubusercontent.com/75827115/148100779-6a8d7e12-98d7-4236-a6f9-07f7d2fd3718.JPG)

**Reply by ff6347** [reply]
Can you please post the whole error message and not a cut off screenshot. The interesting part is not visible here

**Reply by JerreyGlen** [reply]
![error2](https://user-images.githubusercontent.com/75827115/148105863-e976e292-98ce-4a1a-b235-e8a4719d56d5.JPG)
![error1](https://user-images.githubusercontent.com/75827115/148105868-79f7109a-f204-461b-b094-37a5c6c1f7d2.JPG)

**Reply by ff6347** [reply]
Please copy and paste the message. The important part is the bad request message that is cut off on the right hand side.

**Reply by JerreyGlen** [reply]
I have no error message more. it had worked. thanks for all
![stadtpuls](https://user-images.githubusercontent.com/75827115/148127308-f8b83494-0724-4675-951b-a5ac1ae377c3.JPG)

**Reply by ff6347** [reply]
Glad to hear that.

**Reply by CargoBikoMeter** [reply]
Today I had success by sending measurements via my Logstash pipeline.
The following JSON works:
{"measurements":["22.44","1026","47.27","7.5","3.3","3.184"]}

It would be nice, if the following JSON would also work:
{"app_id":"adlerkiez-pm-sensor-01","measurements":["22.2","1025","46.78","8.6","3.7","3.189"]}

I need the app_id inside my Logstash output pipeline to suppress JSON outputs from other devices.

Is it possible to modify the HTTP-API to search only for measurements JSON tags?

**Reply by ff6347** [reply]
Hi Roland,

Yes we can do that. It's a small change. Currently I'm on vacation so it might take 2-3 weeks until this I'd deployed. 

If you can't wait for it you could do a pull request and write some tests for it. The line to start at is this one

https://github.com/technologiestiftung/stadtpuls-api/blob/c35ab7e750a84431ec4fefe465f3e2cb5b7b2135/src/integrations/http.ts#L32

All PRs should go against the staging branch.

**Reply by CargoBikoMeter** [reply]
Hi, today I found the solution in Logstash (https://discuss.elastic.co/t/http-output-for-stadtpuls-com-iot-platform/296777/4) and a modification of the stadtpuls API is not required.

**Reply by Esshahn** [reply]
Great!
